blackjax/util.py

```diff
@@ -142,12 +142,13 @@ def index_pytree(input_pytree: ArrayLikeTree) -> ArrayTree:
 
 def run_inference_algorithm(
     rng_key: PRNGKey,
-    initial_state_or_position: ArrayLikeTree,
+    initial_state: ArrayLikeTree,
```

Comment:
While I agree the initial `try ... except ...` block is not rigor to capture both `initial_state` and `initial_position`, we should not remove it here because an important use case for this top level API is for user to just supply the `initial_position`. Let's do it this way:

```python
def run_inference_algorithm(
    rng_key: PRNGKey,
    inference_algorithm: Union[SamplingAlgorithm, VIAlgorithm],
    num_steps: int,
    initial_state: ArrayLikeTree = None,
    initial_position: ArrayLikeTree = None,
    progress_bar: bool = False,
    transform: Callable = lambda x: x,
):
    ...
    if initial_state is None and initial_position is None:
        raise ...
    if initial_state is not None and initial_position is not None:
        raise ...
    if initial_position is not None:
        rng_key, init_key = split(rng_key, 2)
        inital_state = inference_algorithm.init(initial_state_or_position, init_key)
```

Reply:
OK, I'm open to that. But I'd argue that this isn't ideal. The user has access to `inference_algorithm.init`, so can easily produce `initial_state` from `initial_position` in just one line. Hiding this process inside `run_inference_algorithm` doesn't really make it easier for the user, and creates an awkward `initial_state_or_position` variable along with various errors that have to be raised when both or neither are present. 

tldr: benefit of this feature is very low, added code complexity is high

Reply:
It is a valid point, but then maybe we should change to require the input of `inference_algorithm.step` in the API?

Reply:
That seems reasonable. If you like, I can do that now.

Reply:
OK, I put it in, as you outlined. This required changing the order of arguments, so I went through the code and made sure `run_inference_algorithm` is called with default arguments everywhere.